Inconsistent Client Statistic Value for System Load Average

As you can see [here](https://github.com/hazelcast/hazelcast-nodejs-client/blob/a608c834c761727c841e404fc8ad21484859554e/src/statistics/Statistics.ts#L157), system load average is sent as an array of load averages (last 1st, 5th, 15th minutes). However, this value is different in java client which is a double (only last 15th minute). This array value causes an exception on management center during client statistics parse operation as below;

```
2018-12-27 12:06:46 [qtp1873653341-74] WARN  c.h.w.service.CentralManager - Failed to parse client stats.
com.hazelcast.webmonitor.model.ClientStatsParseException: Failed to parse client stats.
	at com.hazelcast.webmonitor.model.ClientStats.create(ClientStats.java:63)
	at com.hazelcast.webmonitor.model.AllState.addClientStats(AllState.java:328)
	at com.hazelcast.webmonitor.model.AllState.addState(AllState.java:301)
	at com.hazelcast.webmonitor.service.CentralManager.updateAllStateFromLatestState(CentralManager.java:344)
	at com.hazelcast.webmonitor.service.CentralManager.addMemberState(CentralManager.java:288)
	at com.hazelcast.webmonitor.controller.CollectorController.doCollect(CollectorController.java:32)
	at sun.reflect.GeneratedMethodAccessor37.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:133)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:97)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:849)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:760)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:872)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1655)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:208)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1634)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1317)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1219)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:61)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:531)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:762)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:680)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NumberFormatException: For input string: "[2.73583984375"
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.lang.Double.parseDouble(Double.java:538)
	at java.lang.Double.valueOf(Double.java:502)
	at com.hazelcast.webmonitor.model.ClientStatsUtil.getDoubleOrNull(ClientStatsUtil.java:33)
	at com.hazelcast.webmonitor.model.ClientOSStats.create(ClientOSStats.java:36)
	at com.hazelcast.webmonitor.model.ClientStats.create(ClientStats.java:79)
	at com.hazelcast.webmonitor.model.ClientStats.create(ClientStats.java:61)
```

I believe that, there should be a contract at least on member or management center to prevent this kind of data mapping issues. Otherwise, non-java clients should have same format with java especially for os statistics

--- a/src/Util.ts
+++ b/src/Util.ts
@@ -330,3 +330,9 @@ export function DeferredPromise<T>(): Promise.Resolver<T> {
         promise,
     } as Promise.Resolver<T>;
 }
+
+export function getNodejsMajorVersion(): number {
+    const versionString = process.version;
+    const versions = versionString.split('.');
+    return Number.parseInt(versions[0].substr(1));
+}

--- a/src/statistics/Statistics.ts
+++ b/src/statistics/Statistics.ts
@@ -153,15 +153,22 @@ export class Statistics {
         this.registerGauge('os.freeSwapSpaceSize', () => Statistics.EMPTY_STAT_VALUE);
         this.registerGauge('os.maxFileDescriptorCount', () => Statistics.EMPTY_STAT_VALUE);
         this.registerGauge('os.openFileDescriptorCount', () => Statistics.EMPTY_STAT_VALUE);
-        this.registerGauge('os.processCpuTime', () => Statistics.EMPTY_STAT_VALUE);
-        this.registerGauge('os.systemLoadAverage', () => '[' + os.loadavg().toString() + ']');
+        this.registerGauge('os.processCpuTime', () => {
+            // Nodejs 4 does not support this metric. So we do not print an ugly warning for that.
+            if (Util.getNodejsMajorVersion() >= 6) {
+                return process.cpuUsage().user * 1000; // process.cpuUsage returns micoseconds. We convert to nanoseconds.
+            } else {
+                return Statistics.EMPTY_STAT_VALUE;
+            }
+        });
+        this.registerGauge('os.systemLoadAverage', () => os.loadavg()[0]);
         this.registerGauge('os.totalPhysicalMemorySize', () => os.totalmem());
         this.registerGauge('os.totalSwapSpaceSize', () => Statistics.EMPTY_STAT_VALUE);
         this.registerGauge('runtime.availableProcessors', () => os.cpus().length);
         this.registerGauge('runtime.freeMemory', () => Statistics.EMPTY_STAT_VALUE);
         this.registerGauge('runtime.maxMemory', () => Statistics.EMPTY_STAT_VALUE);
         this.registerGauge('runtime.totalMemory', () => process.memoryUsage().heapTotal);
-        this.registerGauge('runtime.uptime', () => process.uptime());
+        this.registerGauge('runtime.uptime', () => process.uptime() * 1000);
         this.registerGauge('runtime.usedMemory', () => process.memoryUsage().heapUsed);
         this.registerGauge('executionService.userExecutorQueueSize', () => Statistics.EMPTY_STAT_VALUE);
     }

--- a/test/statistics/StatisticsTest.js
+++ b/test/statistics/StatisticsTest.js
@@ -14,13 +14,16 @@
  * limitations under the License.
  */
 
+var Statistics = require("../../lib/statistics/Statistics").Statistics;
 var expect = require('chai').expect;
 var BuildInfo = require('../../lib/BuildInfo').BuildInfo;
 
 var RC = require('../RC');
 var Client = require('../../').Client;
-var Util = require('../Util');
+var TestUtil = require('../Util');
+var Util = require('../../lib/Util');
 var Config = require('../../').Config;
+var os = require('os');
 
 describe('Statistics with default period', function () {
 
@@ -29,7 +32,7 @@ describe('Statistics with default period', function () {
     var map;
 
     before(function () {
-        Util.markServerVersionAtLeast(this, null, '3.9.0');
+        TestUtil.markServerVersionAtLeast(this, null, '3.9.0');
         return RC.createCluster(null, null).then(function (res) {
             cluster = res;
         }).then(function () {
@@ -48,7 +51,7 @@ describe('Statistics with default period', function () {
     });
 
     after(function () {
-        Util.markServerVersionAtLeast(this, null, '3.9.0');
+        TestUtil.markServerVersionAtLeast(this, null, '3.9.0');
         client.shutdown();
         return RC.shutdownCluster(cluster.id);
     });
@@ -64,7 +67,7 @@ describe('Statistics with default period', function () {
     });
 
     it('should be enabled via configuration', function () {
-        return Util.promiseWaitMilliseconds(1000).then(function () {
+        return TestUtil.promiseWaitMilliseconds(1000).then(function () {
             return getClientStatisticsFromServer(cluster, client);
         }).then(function (stats) {
             expect(stats).to.not.null;
@@ -73,34 +76,38 @@ describe('Statistics with default period', function () {
     });
 
     it('should contain statistics content', function () {
-        return Util.promiseWaitMilliseconds(1000).then(function () {
+        return TestUtil.promiseWaitMilliseconds(1000).then(function () {
             return getClientStatisticsFromServer(cluster, client);
         }).then(function (stats) {
             expect(stats).to.not.be.null;
-            expect(contains(stats, 'clientName=' + client.getName())).to.be.true;
-            expect(contains(stats, 'lastStatisticsCollectionTime=')).to.be.true;
-            expect(contains(stats, 'enterprise=false')).to.be.true;
-            expect(contains(stats, 'clientType=NodeJS')).to.be.true;
-            expect(contains(stats, 'clientVersion=' + BuildInfo.getClientVersion())).to.be.true;
-
+            expect(extractStringStatValue(stats, 'clientName')).to.equal(client.getName());
+            expect(extractIntStatValue(stats, 'lastStatisticsCollectionTime')).to.be
+                .within(Date.now() - Statistics.PERIOD_SECONDS_DEFAULT_VALUE * 2000, Date.now());
+            expect(extractBooleanStatValue(stats, 'enterprise')).to.be.false;
+            expect(extractStringStatValue(stats, 'clientType')).to.equal('NodeJS');
+            expect(extractStringStatValue(stats, 'clientVersion')).to.equal(BuildInfo.getClientVersion());
             var ownerConnection = client.getClusterService().getOwnerConnection();
-            expect(contains(stats, 'clusterConnectionTimestamp=' + ownerConnection.getStartTime())).to.be.true;
-            expect(contains(stats, 'clientAddress=' + ownerConnection.getLocalAddress().toString())).to.be.true;
-            expect(contains(stats, 'os.committedVirtualMemorySize=')).to.be.true;
-            expect(contains(stats, 'os.freeSwapSpaceSize=')).to.be.true;
-            expect(contains(stats, 'os.maxFileDescriptorCount=')).to.be.true;
-            expect(contains(stats, 'os.openFileDescriptorCount=')).to.be.true;
-            expect(contains(stats, 'os.processCpuTime=')).to.be.true;
-            expect(contains(stats, 'os.systemLoadAverage=')).to.be.true;
-            expect(contains(stats, 'os.totalPhysicalMemorySize=')).to.be.true;
-            expect(contains(stats, 'os.totalSwapSpaceSize=')).to.be.true;
-            expect(contains(stats, 'runtime.availableProcessors=')).to.be.true;
-            expect(contains(stats, 'runtime.freeMemory=')).to.be.true;
-            expect(contains(stats, 'runtime.maxMemory=')).to.be.true;
-            expect(contains(stats, 'runtime.totalMemory=')).to.be.true;
-            expect(contains(stats, 'runtime.uptime=')).to.be.true;
-            expect(contains(stats, 'runtime.usedMemory=')).to.be.true;
-            expect(contains(stats, 'executionService.userExecutorQueueSize=')).to.be.true;
+            expect(extractIntStatValue(stats, 'clusterConnectionTimestamp')).to.equal(ownerConnection.getStartTime());
+            expect(extractStringStatValue(stats, 'clientAddress')).to.equal(ownerConnection.getLocalAddress().toString());
+            expect(extractStringStatValue(stats, 'os.committedVirtualMemorySize')).to.equal('');
+            expect(extractStringStatValue(stats, 'os.freeSwapSpaceSize')).to.equal('');
+            expect(extractStringStatValue(stats, 'os.maxFileDescriptorCount')).to.equal('');
+            expect(extractStringStatValue(stats, 'os.openFileDescriptorCount')).to.equal('');
+            if (Util.getNodejsMajorVersion() >= 6) {
+                expect(extractIntStatValue(stats, 'os.processCpuTime')).to.greaterThan(1000);
+            } else {
+                expect(extractStringStatValue(stats, 'os.processCpuTime')).to.equal('');
+            }
+            expect(extractFloatStatValue(stats, 'os.systemLoadAverage')).to.be.greaterThan(0);
+            expect(extractIntStatValue(stats, 'os.totalPhysicalMemorySize')).to.equal(os.totalmem());
+            expect(extractStringStatValue(stats, 'os.totalSwapSpaceSize')).to.equal('');
+            expect(extractIntStatValue(stats, 'runtime.availableProcessors')).to.equal(os.cpus().length);
+            expect(extractStringStatValue(stats, 'runtime.freeMemory')).to.equal('');
+            expect(extractStringStatValue(stats, 'runtime.maxMemory')).to.equal('');
+            expect(extractIntStatValue(stats, 'runtime.totalMemory')).to.greaterThan(0);
+            expect(extractIntStatValue(stats, 'runtime.uptime')).to.greaterThan(0);
+            expect(extractIntStatValue(stats, 'runtime.usedMemory')).to.greaterThan(0);
+            expect(extractStringStatValue(stats, 'executionService.userExecutorQueueSize')).to.equal('');
         });
     });
 
@@ -110,7 +117,7 @@ describe('Statistics with default period', function () {
         }).then(function () {
             return map.get('key');
         }).then(function () {
-            return Util.promiseWaitMilliseconds(5000);
+            return TestUtil.promiseWaitMilliseconds(5000);
         }).then(function () {
             return getClientStatisticsFromServer(cluster, client);
         }).then(function (stats) {
@@ -133,7 +140,7 @@ describe('Statistics with non-default period', function () {
     var client;
 
     before(function () {
-        Util.markServerVersionAtLeast(this, null, '3.9.0');
+        TestUtil.markServerVersionAtLeast(this, null, '3.9.0');
         return RC.createCluster(null, null).then(function (res) {
             cluster = res;
         }).then(function () {
@@ -149,14 +156,14 @@ describe('Statistics with non-default period', function () {
     });
 
     after(function () {
-        Util.markServerVersionAtLeast(this, null, '3.9.0');
+        TestUtil.markServerVersionAtLeast(this, null, '3.9.0');
         client.shutdown();
         return RC.shutdownCluster(cluster.id);
     });
 
     it('should not change before period', function () {
         var stats1;
-        return Util.promiseWaitMilliseconds(1000).then(function () {
+        return TestUtil.promiseWaitMilliseconds(1000).then(function () {
             return getClientStatisticsFromServer(cluster, client);
         }).then(function (st) {
             stats1 = st;
@@ -168,11 +175,11 @@ describe('Statistics with non-default period', function () {
 
     it('should change after period', function () {
         var stats1;
-        return Util.promiseWaitMilliseconds(1000).then(function () {
+        return TestUtil.promiseWaitMilliseconds(1000).then(function () {
             return getClientStatisticsFromServer(cluster, client);
         }).then(function (st) {
             stats1 = st;
-            return Util.promiseWaitMilliseconds(2000)
+            return TestUtil.promiseWaitMilliseconds(2000)
         }).then(function () {
             return getClientStatisticsFromServer(cluster, client);
         }).then(function (stats2) {
@@ -186,7 +193,7 @@ describe('Statistics with negative period', function () {
     var cluster;
 
     before(function () {
-        Util.markServerVersionAtLeast(this, null, '3.9.0');
+        TestUtil.markServerVersionAtLeast(this, null, '3.9.0');
         return RC.createCluster(null, null).then(function (res) {
             cluster = res;
         }).then(function () {
@@ -202,13 +209,13 @@ describe('Statistics with negative period', function () {
     });
 
     after(function () {
-        Util.markServerVersionAtLeast(this, null, '3.9.0');
+        TestUtil.markServerVersionAtLeast(this, null, '3.9.0');
         client.shutdown();
         return RC.shutdownCluster(cluster.id);
     });
 
     it('should be enabled via configuration', function () {
-        return Util.promiseWaitMilliseconds(1000).then(function () {
+        return TestUtil.promiseWaitMilliseconds(1000).then(function () {
             return getClientStatisticsFromServer(cluster, client);
         }).then(function (stats) {
             expect(stats).to.not.equal('');
@@ -232,4 +239,22 @@ function getClientStatisticsFromServer(cluster, client) {
         }
         return null;
     });
+}
+
+function extractStringStatValue(stats, statName) {
+    var re = new RegExp(statName + '=(.*?)(?:,|$)');
+    var matches = stats.match(re);
+    return matches[1];
+}
+
+function extractFloatStatValue(stats, statName) {
+    return Number.parseFloat(extractStringStatValue(stats, statName));
+}
+
+function extractBooleanStatValue(stats, statName) {
+    return 'true' === extractStringStatValue(stats, statName);
+}
+
+function extractIntStatValue(stats, statName) {
+    return Number.parseInt(extractStringStatValue(stats, statName));
 }

